### PR TITLE
Birthday attack follow: Update additional server configs with tls config containing restricted set of cipher suites

### PR DESCRIPTION
**Pull request by benjaminapetersen**
Follow-on to #3389 

[Full list of related PRs](https://github.com/openshift/console/pulls?utf8=%E2%9C%93&q=is%3Apr+Birthday+attack+against+64-bit+block)

Will need a cherry-pick including both commits for backport.

TODO:
- update `test-ciphers.sh` to validate that this does fix the problem.

### CODE REVIEW

File: test/scripts/test-ciphers-from-pod.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+echo "connecting to ${SERVER}"
+
+
+# ensure we ignore weak ciphers
+# CBC (cipher block chaining) are no longer reliable and should not be used
+# CBC ciphers use an IV (initialization vector) and a chaining mechanism.
+# The chaining mechanism means that a single bit error in a ciphertext block
+# will invalidate all previous blocks.  The chaining was good in that it hides
+# plaintext patterns, but is inferior to other cipher modes.
+INVALID_CIPHER_SAMPLE=(
```
Comment (benjaminapetersen) [inline test/scripts/test-ciphers-from-pod.sh]: These are correctly denied.

File: test/scripts/test-ciphers-from-pod.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  then
+    echo "invalid cipher suite used to connect to console (${CIPHER})"
+    exit 1
+  else
+    echo "invalid cipher was correctly denied (${CIPHER})"
+  fi
+done
+
+
+# CIPHER=ECDHE-ECDSA-CHACHA20-POLY1305 # DENIED
```
Comment (benjaminapetersen) [inline test/scripts/test-ciphers-from-pod.sh]: However, these are also denied :)
Working on figuring out what is up with this.

Comment (benjaminapetersen) [reply test/scripts/test-ciphers-from-pod.sh]: @spadgett curious if you have a thought.

File: test-ciphers.sh [hunk 3]
```diff
@@ -1,58 +1,61 @@
...
+cat <<EOF | oc create -f -
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: openshift-console
+  name: "${POD_NAME}"
+spec:
+  restartPolicy: Never
+  containers:
+  - name: "${POD_NAME}"
```
Comment (benjaminapetersen) [inline test-ciphers.sh]: Tried all of these and more:

```
# image: frapsoft/openssl
    # image: fedora:latest
    # image: ubuntu
    # image: quay.io/coreos/tectonic-console-builder:v19
```
Had to build my own to get the tools needed.

File: test/Dockerfile.ciphertest [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# FROM ubuntu:latest
```
Comment (benjaminapetersen) [inline test/Dockerfile.ciphertest]: Didn't really want to create my own test image, but couldn't find the correct mix of tools otherwise.

File: test/test-ciphers-from-pod.sh [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+)
+
+echo "checking for console"
+curl --head "${SERVER}"
+
+for CIPHER in "${INVALID_CIPHER_SAMPLE[@]}"
+do
+  RESULT=$(openssl s_client -connect "${SERVER}" -cipher "${CIPHER}" 2>&1)
+  # note that this swallows errors if openssl doesn't exist in the container
+  if [[ $? -eq 0 ]]
```
Comment (benjaminapetersen) [inline test/test-ciphers-from-pod.sh]: This is bad, I used to check for specific strings in the output, which might be better.  This will swallow errors if openssl isn't installed.  On the flip side, with the `VALID_CIPHER_SAMPLE` tests below, these also fail.  The reason is just unclear (the kind of reasons that suck in CI when you forget past pains).

File: test/Dockerfile.ciphertest (hunk 4 above)
Comment (benjaminapetersen) [reply test/Dockerfile.ciphertest]:
```
 docker build -f Dockerfile.ciphertest  -t quay.io/benjaminapetersen/console-ciphertest:latest .
```
and exists here https://quay.io/repository/benjaminapetersen/console-ciphertest

File: test-ciphers.sh [hunk 6]
```diff
@@ -1,56 +1,66 @@
...
-  if openssl s_client -connect "${SERVER}" -cipher "${CIPHER}" -CAfile /tmp/default-ingress-cert-file.txt 2>&1
-  then
-    echo "invalid cipher suite used to connect to console (${CIPHER})"  
-    exit 1  
-  else    
-    echo "invalid cipher was correctly denied (${CIPHER})"
-  fi
-done
+echo "server ciphers incorrectly handled"
+exit 1
```
Comment (benjaminapetersen) [inline test-ciphers.sh]: Final output for this test looks like:

```bash
invalid cipher was correctly denied (DES-CBC3-SHA)
invalid cipher was correctly denied (RSA-AES-128-CBC-SHA256)
invalid cipher was correctly denied (ECDHE-RSA-3DES-EDE-CBC-SHA)
invalid cipher was correctly denied (RSA-3DES-EDE-CBC-SHA)
valid cipher was correctly accepted (ECDHE-RSA-AES128-GCM-SHA256)
valid cipher was correctly accepted (ECDHE-RSA-AES256-GCM-SHA384)
valid cipher was correctly accepted (ECDHE-ECDSA-CHACHA20-POLY1305)
valid cipher was correctly accepted (ECDHE-RSA-CHACHA20-POLY1305)
success
server ciphers correctly handled
```

File: test-ciphers.sh [hunk 7]
```diff
@@ -1,56 +1,66 @@
...
+      name: "${CONFIGMAP_NAME}"
+      defaultMode: 0755
+EOF
+
+# wait to ensure the tests run
+sleep 10
+
+# check the logs to see that ciphers were handled as expected
+LOGS=$(oc logs "${POD_NAME}")
+echo "${LOGS}"
```
Comment (benjaminapetersen) [inline test-ciphers.sh]: dumping the logs from the pod to make debugging easier if this ever breaks for some reason

File: test-ciphers.sh (hunk 3 above)
Comment (benjaminapetersen) [reply test-ciphers.sh]: Unfortunately messy, but I included the Dockerfile below.

File: cmd/bridge/main.go [hunk 8]
```diff
@@ -535,6 +535,10 @@ func main() {
 		Handler: srv.HTTPHandler(),
 		// Disable HTTP/2, which breaks WebSockets.
 		TLSNextProto: make(map[string]func(*http.Server, *tls.Conn, http.Handler)),
+		TLSConfig: &tls.Config{
+			RootCAs:      x509.NewCertPool(),
```
Comment (benjaminapetersen) [inline cmd/bridge/main.go]: Shouldn't need the `RootCAs:      x509.NewCertPool(),` here, removing.

File: test-ciphers.sh [hunk 9]
```diff
@@ -1,56 +1,70 @@
...
+    - name: "${CONFIGMAP_NAME}"
+      mountPath: ${FILE_PATH}
+  volumes:
+  - name: "${CONFIGMAP_NAME}"
+    configMap:
+      name: "${CONFIGMAP_NAME}"
+      defaultMode: 0755
+EOF
+
+# wait to ensure the tests run
```
Comment (benjaminapetersen) [inline test-ciphers.sh]: I think something like 
```
oc wait --for=condition=complete --timeout=32s "pod/${podname}"
```
would be better than sleep. will test to confirm.

File: test-ciphers.sh [hunk 10]
```diff
@@ -1,56 +1,70 @@
...
+sleep 10
```
Comment (spadgett) [inline test-ciphers.sh]: Don't put in an arbitrary wait. You should be able to use `oc wait`

File: test-ciphers.sh [hunk 11]
```diff
@@ -1,56 +1,70 @@
...
+
+# check the logs to see that ciphers were handled as expected
+LOGS=$(oc logs "${POD_NAME}")
+echo "${LOGS}"
+
+if [[ "${LOGS}" == *"success"* ]]; then
```
Comment (spadgett) [inline test-ciphers.sh]: Let's not check for text in the logs. I'd just check the exit status of the container.

File: test-ciphers.sh [hunk 12]
```diff
@@ -1,56 +1,70 @@
...
+# put the test script in a configmap in the console namespace
+oc create configmap "${CONFIGMAP_NAME}" --from-file="./test/${FILE_NAME}"
+
+# creates a POD to run the test script in the console namespace
+# passes $CONSOLE_POD_IP as $SERVER environment variable
+cat <<EOF | oc create -f -
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: openshift-console
```
Comment (spadgett) [inline test-ciphers.sh]: Does this pod need to live in the openshift-console namespace? I feel like it would be better to create and delete a test namespace, then there's no clean up to worry about.

File: test-ciphers.sh [hunk 13]
```diff
@@ -1,56 +1,70 @@
-#!/usr/bin/env bash
+#!/bin/bash
```
Comment (spadgett) [inline test-ciphers.sh]: This is less portable the `#!/usr/bin/env bash` and inconsistent with other console scripts. I'd undo this change.

File: test/test-ciphers-from-pod.sh [hunk 14]
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
```
Comment (spadgett) [inline test/test-ciphers-from-pod.sh]:
```suggestion
#!/usr/bin/env bash
```

File: test-ciphers.sh [hunk 15]
```diff
@@ -1,56 +1,70 @@
...
-then
-  echo "no console_url, are you connected to a cluster?"
-  exit 1
+set -o errexit
+set -o nounset
+set -o pipefail
+
+oc project openshift-console
+
+CONSOLEPOD=$(oc get pods --template '{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}' | grep --max-count 1 console)
```
Comment (spadgett) [inline test-ciphers.sh]:
```suggestion
CONSOLEPOD=$(oc get pod -l app=console -o jsonpath="{.items[0].metadata.name}" -n openshift-console)
```

File: test/Dockerfile.ciphertest (hunk 4 above)
Comment (spadgett) [inline test/Dockerfile.ciphertest]: Remove commented-out code

Comment (spadgett) [reply test/Dockerfile.ciphertest]: I'm not sure we want to host this on your personal quay account. We could potentially use OpenShift builds to build the image.

**Comment by benjaminapetersen** [comment]
@spadgett agree about hosting (I don't want it on my personal quay either), not sure how much work it is to setup OpenShift builds.

File: test-ciphers.sh (hunk 13 above)
Comment (benjaminapetersen) [reply test-ciphers.sh]: will do.

File: test-ciphers.sh (hunk 15 above)
Comment (benjaminapetersen) [reply test-ciphers.sh]: 👍

File: test-ciphers.sh (hunk 12 above)
Comment (benjaminapetersen) [reply test-ciphers.sh]: works for me.

File: test-ciphers.sh (hunk 10 above)
Comment (benjaminapetersen) [reply test-ciphers.sh]: Agree!

File: test-ciphers.sh (hunk 11 above)
Comment (benjaminapetersen) [reply test-ciphers.sh]: Agree, something better would be good here.  I don't recall if that's in the `pod.status.containerStatuses`, will take a look once I get a cluster up.

File: test/Dockerfile.ciphertest (hunk 4 above)
Comment (benjaminapetersen) [reply test/Dockerfile.ciphertest]: +1

File: test/test-ciphers-from-pod.sh [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+set -o errexit
```
Comment (benjaminapetersen) [inline test/test-ciphers-from-pod.sh]: Can't use 

```
set -o errexit
```
As some of the test requires an error. I'm not sure if there is a better approach.

File: test-ciphers.sh (hunk 11 above)
Comment (spadgett) [reply test-ciphers.sh]: Yes, you can get the exit code from the container from container statuses

File: test-ciphers.sh [hunk 17]
```diff
@@ -1,56 +1,74 @@
...
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: openshift-console-test
+  name: "${POD_NAME}"
+spec:
+  restartPolicy: Never
+  containers:
+  - name: "${POD_NAME}"
+    image: quay.io/benjaminapetersen/console-ciphertest:latest
```
Comment (spadgett) [inline test-ciphers.sh]: @benjaminapetersen You should be able to use our builder image, which already has openssl

```suggestion
    image: quay.io/coreos/tectonic-console-builder:v19
```

**Comment by benjaminapetersen** [comment]
Updated, eliminated the cipher tests as we've decided it's not really worth the effort.

**Comment by spadgett** [comment]
/hold
for #5100
